File: legacy/src/main/java/io/github/notenoughupdates/moulconfig/internal/ForgeFontRenderer.java
package io.github.notenoughupdates.moulconfig.internal;

import io.github.notenoughupdates.moulconfig.common.IFontRenderer;
import io.github.notenoughupdates.moulconfig.common.text.StructuredText;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.GuiUtilRenderComponents;
import net.minecraft.util.IChatComponent;

import java.util.ArrayList;
import java.util.List;

public class ForgeFontRenderer implements IFontRenderer {
    public final FontRenderer font;

    public ForgeFontRenderer(FontRenderer font) {
        this.font = font;
    }

    @Override
    public int getStringWidth(StructuredText string) {
        return font.getStringWidth(StructuredTextImpl.unwrap(string).getFormattedText());
    }

    @Override
    public int getHeight() {
        return font.FONT_HEIGHT;
    }

    @Override
    public int getStringWidth(String string) {
        return font.getStringWidth(string);
    }

    @Override
    public int getCharWidth(char c) {
        return font.getCharWidth(c);
    }

    @Override
    public List<StructuredText> splitText(StructuredText text, int width) {
        List<IChatComponent> components = GuiUtilRenderComponents.splitText(StructuredTextImpl.unwrap(text), width, font, false, false);
        List<StructuredText> result = new ArrayList<>();
        for (IChatComponent component : components) {
            result.add(StructuredTextImpl.wrap(component));
        }
        return result;
    }

    @Override
    public String trimStringToWidth(String string, int width, boolean reverse) {
        return font.trimStringToWidth(string, width, reverse);
    }
}

File: legacy/src/main/java/io/github/notenoughupdates/moulconfig/internal/ForgeFontRenderer.kt
package io.github.notenoughupdates.moulconfig.internal

import io.github.notenoughupdates.moulconfig.common.IFontRenderer
import io.github.notenoughupdates.moulconfig.common.text.StructuredText
import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.gui.GuiUtilRenderComponents
import java.util.regex.Pattern


class ForgeFontRenderer(val font: FontRenderer) : IFontRenderer {
    override val height: Int
        get() = font.FONT_HEIGHT

    override fun getStringWidth(string: StructuredText): Int {
        return font.getStringWidth(StructuredTextImpl.unwrap(string).formattedText)
    }

    override fun getStringWidth(string: String): Int {
        return font.getStringWidth(string)
    }

    override fun getCharWidth(char: Char): Int {
        return font.getCharWidth(char)
    }

    override fun splitText(text: StructuredText, width: Int): List<StructuredText> {
        val iChatComponents =
            GuiUtilRenderComponents.splitText(StructuredTextImpl.unwrap(text), width, font, false, false)
        return iChatComponents.map { StructuredTextImpl.wrap(it) }
    }

    override fun splitLines(text: StructuredText): List<StructuredText> {
        return splitText(text, Integer.MAX_VALUE)
    }

    override fun trimStringToWidth(string: String, width: Int, reverse: Boolean): String {
        return font.trimStringToWidth(string, width, reverse)
    }
}
